android/java/org/chromium/chrome/browser/search_engines/BraveTemplateUrlServiceFactory.java

```diff
@@ -0,0 +1,97 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.chromium.chrome.browser.search_engines;
+
+import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
+
+import org.chromium.base.Log;
+import org.chromium.base.ThreadUtils;
+import org.chromium.base.annotations.NativeMethods;
+import org.chromium.chrome.browser.profiles.Profile;
+import org.chromium.components.search_engines.TemplateUrlService;
+
+import  java.lang.reflect.InvocationTargetException;
+
+public class BraveTemplateUrlServiceFactory {
+    private static final String TAG = "BraveTemplateUrlServiceFactory";
+
+    private static TemplateUrlService sNormalTemplateUrlService;
+    private static @Nullable Profile sProfile;
+
+    private BraveTemplateUrlServiceFactory() {}
+
+    /**
+     * @return The singleton instance of {@link TemplateUrlService} for profile, based on current active tab.
+     * All TemplateUrlServiceFactory.get() calls are redirected here.
+     */
+    public static TemplateUrlService get() {
+        ThreadUtils.assertOnUiThread();
+
+        if (sProfile != null) return get(sProfile);
+
+        Profile profile = null;
+
+        // Cannot import BraveActivity here due to circular dependency, need reflection all the way
+        Class<?> braveActivityClazz = null;
+        try{
+            braveActivityClazz = Class.forName( "org.chromium.chrome.browser.app.BraveActivity" );
+        } catch (ClassNotFoundException  e) {
+            Log.w(TAG, "Failed to get BraveActivity class: " + e.getMessage());
+        }
+
+        if (braveActivityClazz != null) {
+            try {
+                Object activity = braveActivityClazz.getMethod("getBraveActivity").invoke(null);
+                if (activity != null) {
+                    profile = (Profile) braveActivityClazz.getMethod("getCurrentProfile").invoke(activity);
+                }
+            } catch (NoSuchMethodException | SecurityException | IllegalAccessException | IllegalArgumentException | InvocationTargetException | ClassCastException e) {
+                Log.w(TAG, "Failed to get current active profile: " + e.getMessage());
+            }
+        }
+
+        if (profile == null){
```

Comment:
nit: lint will fail here, `) {`

Reply:
I usually fix formatting problems last (lint doesn't work locally for me so I need to rely on CI for it), but thanks for pointing it out!